plugins/saveactions.c

```diff
@@ -296,11 +341,341 @@ static void instantsave_document_new_cb(GObject *obj, GeanyDocument *doc, gpoint
+		gboolean file_exists;
+
+		next_file_name = g_strdup_printf("%s%d%s", PERSISTENT_UNTITLED_DOC_FILE_NAME_PREFIX, i, extension_postfix);
+		next_file_path = g_strdup_printf("%s%c%s", persistent_docs_target_dir, G_DIR_SEPARATOR, next_file_name);
+
+		file_exists = g_file_test(next_file_path, G_FILE_TEST_EXISTS);
+
+		g_free(next_file_path);
+
+		if (file_exists)
+		{
+			g_free(next_file_name);
+		}
+		else
+		{
+			if (strlen(extension_postfix))
+				g_free(extension_postfix);
+
+			return next_file_name;
+		}
+	}
+
+	g_free(extension_postfix);
+
+	return NULL;
+}
+
+
+static void persistent_doc_new_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	if (doc->file_name == NULL)
+	{
+		gchar *files_dir_utf8, *new_file_name_utf8, *new_file_path_utf8;
+
+		if (EMPTY(persistent_docs_target_dir))
+		{
+			dialogs_show_msgbox(GTK_MESSAGE_ERROR,
+				_("Persistent untitled document directory does not exist or is not writable."));
+			return;
+		}
+
+		new_file_name_utf8 = create_new_persistent_doc_file_name(doc);
+
+		document_set_filetype(doc, filetypes_lookup_by_name(untitled_doc_default_ft));
+
+		if (new_file_name_utf8 == NULL)
+		{
+			return;
+		}
+
+		files_dir_utf8 = utils_get_utf8_from_locale(persistent_docs_target_dir);
+
+		new_file_path_utf8 = g_strconcat(files_dir_utf8,
+			G_DIR_SEPARATOR_S, new_file_name_utf8, NULL);
+
+		document_save_file_as(doc, new_file_path_utf8);
+
+		g_free(new_file_name_utf8);
+		g_free(files_dir_utf8);
+		g_free(new_file_path_utf8);
+	}
+}
+
+
+static gboolean document_is_empty(GeanyDocument *doc)
+{
+	/* if total text length is 1 while line count is 2 - then the only character of whole text is a linebreak,
+	which is how completely empty document saved by Geany to disk looks like */
+	return sci_get_length(doc->editor->sci) == 0
+		|| (sci_get_length(doc->editor->sci) == 1 && sci_get_line_count(doc->editor->sci) == 2);
+}
+
+
+/* Open document in idle - useful when trying to re-open document from 'document-close' callback
+ *
+ * @param pointer to document path (locale-encoded)
+ *
+ * @return always FALSE = Just a one shot execution
+ *
+ */
+static gboolean open_document_once_idle(gpointer p_locale_file_path)
+{
+	gchar *locale_file_path = (gchar *) p_locale_file_path;
+
+	document_open_file(locale_file_path, FALSE, NULL, NULL);
+
+	g_free(locale_file_path);
+
+	return FALSE;
+}
+
+
+static gint run_dialog_for_persistent_doc_tab_closing(const gchar *msg, const gchar *msg2)
+{
+	GtkWidget *dialog, *button;
+	gint ret;
+
+	dialog = gtk_message_dialog_new(GTK_WINDOW(geany->main_widgets->window), GTK_DIALOG_DESTROY_WITH_PARENT,
+			GTK_MESSAGE_QUESTION, GTK_BUTTONS_NONE, "%s", msg);
+	gtk_window_set_title(GTK_WINDOW(dialog), _("Question"));
+	gtk_message_dialog_format_secondary_text(GTK_MESSAGE_DIALOG(dialog), "%s", msg2);
+	gtk_dialog_add_button(GTK_DIALOG(dialog), GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL);
+
+	button = ui_button_new_with_image(GTK_STOCK_CLEAR, _("_Don't save"));
```

Comment:
I'm  a little worried that "don't save" doesn't sound as destructive as it should so users might lose their work.

What about calling this "_Don't save (discard)"?

Reply:
done